src/transport/SessionHandle.h

```diff
@@ -29,27 +31,29 @@ class Session;
 class SessionHolder;
 
 /** @brief
- *    Non-copyable session reference. All SessionHandles are created within SessionManager. SessionHandle is not
- *    reference *counted, hence it is not allowed to store SessionHandle anywhere except for function arguments and
- *    return values. SessionHandle is short-lived as it is only available as stack variable, so it is never dangling. */
-class SessionHandle
+ *    A non-copyable version of SessionSharedPtr that is always guaranteed to point to a valid Session. This is always
+ *    created within SessionManager.
+ *
+ *    Since this is a strong reference, this should be short-lived and only used as a stack variable.
+ */
+class SessionHandle : protected SessionSharedPtr
```

Comment:
`SessionHandle` and `SessionSharedPtr` are basically same thing exposing different API. `SessionHandle` is immutable, `SessionSharedPtr` is mutable. Instead of let `SessionHandle` derive from `SessionSharedPtr`, I suggest we do it in the opposite way, use `Optional<SessionHandle>` as a field of `SessionSharedPtr`.

In such implemetation, the SessionHandle is a word smaller, and all APIs of `SessionSharedPtr` should target `SessionHandle` not naked session.

The only problem is that `SessionHandle` is should be copy-able by `SessionSharedPtr`, which we can resolve by using a private copy constructor and a friend declare.

The goal of this class is not non-copyable, but prevent from creating this object in heap, to ensure that the object is ephemeral. Generally C++ can't do that, it is just a coding convention.